src/QuestGoal.ts

```diff
@@ -47,7 +47,7 @@ export class QuestGoal extends EventEmitter {
 			config
 		);
 		this.quest = quest;
-		this.state = {};
+		this.state = {} as TState;
```

Comment:
Creating it empty may create problems in other parts of the code base. After all, TState expects a non empty Record<string, unknown>.

Can we initialise TState with an initial string/unknown pair?

Reply:
It doesn't cause issues while building as long as this type assertion exists; do you have suggestions as to an initial non-empty state that makes sense to use here besides an empty object? 

One possibility might be to add an additional optional param that allows one to set the initial state, _however_ this class is mostly based on data coming from the data source rather than dynamically created Quests/QuestGoals in TypeScript code.

The issues I ran into that led into this PR came along when I was extending QuestGoal and using TypeScript. This initialization of state as `{}` did not cause issues, however the typing of state as `object` was problematic.

Reply:
Here's a gist of the file I was converting to TS and I've pointed out where the previous typing caused an error, you can also see how I used an interface to define a custom TState:
https://gist.github.com/seanohue/103cc4b72c655e76116ce27e1238c8eb

Reply:
https://www.typescriptlang.org/play?#code/MYewdgzgLgBFCm0BcMBK9QCcAmAeamAlmAOYA0MArmANZggDuYAfDALwwDeAvgNwBQ-BNAB0ABxBQAhlBDs4mSvAHCoIgLaIIUkvHkByCdNkxCEBUv0DBoSLHVSAngCN4ACSnr5qkQAtPgkA

Here is a playground using a `Record<string, unknown>` that is instantiated as `{}`. You can see that TS does not error, and that it is somewhat permissive in that it lets us set arbitrary props on the record. However, note that any values we get from the record are typed as `unknown` unless we use a type assertion.

Reply:
I think I understand your question better after re-reading the way you phrased it in Slack. 

The types/type assertion here might cause issues if, at the quest data layer, someone wanted to describe quest goal state as something other than an object-like type. For example, an array of objects or a boolean, or a number, etc. This would go against how Ranvier expects quest data to be shaped and so they would have to work around it. TypeScript does give us an error that hints at this if we do not use the type assertion:

<img width="960" alt="Screen Shot 2021-05-30 at 1 39 35 PM" src="https://user-images.githubusercontent.com/8154105/120114405-0e361500-c14d-11eb-95eb-dfdc799d498e.png">

Reply:
I don't know the code base in detail, but the issue here is that you're playing games with the typing system and you allow for a possible mismatch between the code and the types.

Let me explain why:

TState is defined in the codebase as Record<string, unknown>. This type cannot be empty. That means all the code (non only the typing system) should be coded expecting the record to contain data. Adding an empty {} is equal to assigning an undefined (or NULL in C) and this may create problems in other parts of the code.

The best solution would be to follow the old OOP rule of allocate/define and initialise during object creation, but never allocate/define in the constructor and initialise at a later point, because you may have bugs in the time between allocation/definition and initialisation.

If there's no way to create a valid initial state for Record<string, unknown> without huge code reorg, the best way is to define it as Record<string, unknown>|undefined (and code safeguards for undefined) or alternatively use Partial<Record<string, unknown>> and again code assuming the Record may be empty.

I hope this is clear.

Reply:
That makes sense, I like the solution of `Record<string, unknown> | undefined` or the use of `Partial`. 

I suspect creating a valid starting state would require code reorg _and_ would be used in every class extending `QuestGoal`, so that would potentially be a breaking change, something like the following:

```
// current constructor in a class extending QuestGoal:
super(quest, config, player);

this.state = {
  found: false,
  delivered: false
};

// future constructor in the same class after updates:
const initialState = {
  found: false,
  delivered: false,
 };
 
 super(quest, config, player, state);
```

so, this would be a better practice (perhaps) but would possibly break existing extensions of QuestGoal unless we continue gaming the type system.

Reply:
So what I've been doing in my game so far without much trouble is:

Define goaltstate in QuestGoal as:
```export interface IQuestGoalState {
	count?: number;
	[key: string]: any;
}```

Inside of each extended QuestGoal, i'll extend the IQuestGoalConfig and the IQuestGoalState (if needed)

```export interface IKillGoalConfig extends IQuestGoalConfig {
	title: string;
	targets: EntityReference[];
	count: number;
}```

Really only need to handle the state/config of the QuestGoal inside of the specific extended QuestGoal itself. Just use the getProgess() or display() to determine the percent/titles for use in other places. Haven't had to use those extended types in any other files than their extended classes.

Maybe not the best approach but has worked so far in my use cases.

I guess my favorite way of handling it would be having QuestGoal taking any as state then defining that state inside of the extending Goal classes.

Reply:
I like using the generics because it allows for either your way of handling it (@nelsonsbrian) or handling it by passing the state's interface into `QuestGoal`, this way you don't have to use a type assertion each time it is referenced, and it is also compatible with your method of handling things. `Record<string, unknown>` is definitely less permissive than `any`, though, which I'm sure has pros and cons